## Development and Application of Laminar-Specific Gene Expression Meta-analysis of the Human Neocortex
[<img src="https://img.shields.io/badge/Supported%20by-%20CONP%2FPCNO-red?link=https://conp.ca/">](https://conp.ca/)
<!---[![DOI:<your number>](http://img.shields.io/badge/DOI-<your number>-<colour hexcode>.svg)](<doi link>)--->

Welcome to the source code repository for both LaminaRGeneVis and the thesis, "Development and Application of Laminar-Specific Gene Expression Meta-analysis of the Human Neocortex." This contains the code necessary to perform all analyses and figure generation for the thesis, as well as the supplementary analyses and figures found in the publication.

## Abstract
Application of RNA sequencing (RNA-seq) has enabled the characterization of genome-wide gene expression in the human brain, including distinct layers of the neocortex. To advance our understanding of cortical architecture and layer-specific differences in gene expression, we performed a meta-analysis of multiple transcriptomic datasets of gene expression of varying techniques in neurotypical human brains. We observed differences in similarity between bulk-tissue and single-nuclei RNA-seq data, as well as layer-specific enrichment of gene expression. Using the techniques used for the meta-analysis, we developed a web application, LaminaRGeneVis, which allows users to perform similar analyses as performed here. Our results suggest that there is specific enrichment of genes that are relatively consistent across different types of data. These findings and the resulting application provide researchers a platform and reference point to better understand layer-specific gene expression in the human neocortex.

## Publication stemming from thesis:
Kim, E. H, Howard, D., Chen, Y., Tripathy, S. J. and French, L. (2021). LaminaRGeneVis: a tool to visualize gene expression across the laminar architecture of the human neocortex. In submission, Frontiers of Neuroinformatics. Available on [BioRxiv](https://www.biorxiv.org/content/10.1101/2021.05.12.443851v1?rss=1)

## Source Data:
The data for this application has been sourced from these following studies and institutions:

1. [He et al. (2017): ](https://pubmed.ncbi.nlm.nih.gov/28414332/) 
  * This study assayed the whole genome using high-throughput RNA-seq in samples from the DLPFC.
2. [Maynard et al. (2021): ](https://www.nature.com/articles/s41593-020-00787-0) 
  * This study assayed the whole genome through the Visium Platform (10X Genomics) in samples from the DLPFC.
3. [Allen Institute for Brain Science (AIBS): Cell-Type Database](https://portal.brain-map.org/atlases-and-data/rnaseq/human-multiple-cortical-areas-smart-seq)
  * This dataset contains multiple cortical regions and assays the whole genome across roughly 49,000 single-cell nuclei.

## Support
This work was supported in part by funding provided NSERC, and by [Brain Canada](https://braincanada.ca/), in partnership with [Health Canada](https://www.canada.ca/en/health-canada.html), for the [Canadian Open Neuroscience Platform initiative](https://conp.ca/).


[<img src=”https://conp.ca/wp-content/uploads/elementor/thumbs/logo-2-o5e91uhlc138896v1b03o2dg8nwvxyv3pssdrkjv5a.png”>](https://conp.ca/)